src/Bicep.Core.UnitTests/Diagnostics/LinterRuleTests/UseResourceIdFunctionsRuleTests.cs

```diff
@@ -0,0 +1,1302 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+using Bicep.Core.Analyzers.Linter.Rules;
+using Microsoft.VisualStudio.TestTools.UnitTesting;
+
+namespace Bicep.Core.UnitTests.Diagnostics.LinterRuleTests
+{
+    [TestClass]
+    public class UseResourceIdFunctionsRuleTests : LinterRuleTestsBase
+    {
+        private static void CompileAndTest(string bicep, params string[] expectedMessages)
+        {
+            AssertLinterRuleDiagnostics(UseResourceIdFunctionsRule.Code, bicep, expectedMessages, new Options(OnCompileErrors.IncludeErrors, IncludePosition.LineNumberAndColumn));
+        }
+
+        [DataRow(@"
+            @metadata({ Description: 'The name of the Virtual Network to Create' })
+            param virtualNetworkName string
+
+            @metadata({ Description: 'The address range of the new VNET in CIDR format' })
+            param virtualNetworkAddressRange string = '10.0.0.0/16'
+
+            @metadata({ Description: 'The name of the subnet created in the new VNET' })
+            param subnetName string
+
+            @metadata({ Description: 'The address range of the subnet created in the new VNET' })
+            param subnetRange string = '10.0.0.0/24'
+
+            @metadata({ Description: 'The DNS address(es) of the DNS Server(s) used by the VNET' })
+            param DNSServerAddress array
+
+            @metadata({ Description: 'The name of an existing NSG to associate with the subnet' })
+            param NSGName string
+            param location string
+
+            resource virtualNetworkName_resource 'Microsoft.Network/virtualNetworks@2020-05-01' = {
+              name: virtualNetworkName
+              location: location
+              properties: {
+                addressSpace: {
+                  addressPrefixes: [
+                    virtualNetworkAddressRange
+                  ]
+                }
+                dhcpOptions: {
+                  dnsServers: DNSServerAddress
+                }
+                subnets: [
+                  {
+                    name: subnetName
+                    properties: {
+                      addressPrefix: subnetRange
+                      networkSecurityGroup: {
+                        id: '/subscriptions/${subscription().subscriptionId}/resourceGroups/${resourceGroup().name}/providers/Microsoft.Network/networkSecurityGroups/${NSGName}'
+                      }
+                    }
+                  }
+                ]
+              }
+            }",
+            new object[]
+            {
+                // TTK result:      
+                // Property: "id" must use one of the following expressions for an resourceId property:                            
```

Comment:
nit: Format(I see unwanted whitespaces in VS editor) and update description.

Reply:
What do you mean by description?

Reply:
Nvm, looks good. Thanks!